python/lsst/ts/ofc/controllers/zernike_controller.py

```diff
@@ -0,0 +1,356 @@
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import logging
+
+import galsim
+import numpy as np
+
+from .. import OFCData
+from ..utils.ofc_data_helpers import get_intrinsic_zernikes
+
+__all__ = ["ZernikeController"]
+
+
+class ZernikeController:
+    """Zernike Controller.
+
+    This class is mainly used to determine the zernike step (`zk`)
+    at time `k+1` based on the wavefront error (`yk`) at time
+    `k`.
+
+    Parameters
+    ----------
+    ofc_data : `OFCData`
+        OFC Data.
+    log : `logging.Logger` or `None`
+        Optional logging class to be used for logging operations. If `None`,
+        creates a new logger.
+
+    Attributes
+    ----------
+    zk_state : `np.array`
+        State of telescope in the basis of degrees of freedom.
+    zk_state0 : `np.array`
+        Initial state of telescope in the basis of degrees of freedom.
+    integral : `np.array`
+        Integral of the error.
+    kp : `float`
+        Proportional gain.
+    ki : `float`
+        Integral gain.
+    kd : `float`
+        Derivative gain.
+    log : `logging.Logger`
+        Logger class used for logging operations.
+    ofc_data : `OFCData`
+        OFC data.
+    previous_error : `np.array`
+        Previous error.
+    pssn_data : `dict`
+        PSSN data.
+    setpoint : `np.array`
+        Setpoint for the PID controller.
+    """
+
+    # Eta in FWHM calculation.
+    ETA = 1.086
+
+    # FWHM in atmosphere.
+    FWHM_ATM = 0.6
+
+    def __init__(
+        self, ofc_data: OFCData, num_sensors: int = 4, log: logging.Logger | None = None
+    ) -> None:
+        if log is None:
+            self.log = logging.getLogger(type(self).__name__)
+        else:
+            self.log = log.getChild(type(self).__name__)
+
+        # Set OFC data
+        self.ofc_data = ofc_data
+
+        # Set gains and setpoint
+        self._kp = self.ofc_data.zk_controller["kp"]
+        self._ki = self.ofc_data.zk_controller["ki"]
+        self._kd = self.ofc_data.zk_controller["kd"]
+        self._derivative_filter_coeff = self.ofc_data.zk_controller[
+            "derivative_filter_coeff"
+        ]
+        self.setpoint = np.array(self.ofc_data.zk_controller["setpoint"])
+
+        # Set initial state
+        num_zk = self.ofc_data.znmax - self.ofc_data.znmin + 1
+        self.zk_state0 = np.zeros((num_sensors, num_zk))
+        self.zk_state = self.zk_state0.copy()
+
+        # Initialize previous error and integral
+        self.previous_error = self.setpoint - self.zk_state0
+        self.integral = self.previous_error
+        self.filtered_derivative = np.zeros((num_sensors, num_zk))
+
+    @property
```

Comment:
Why are you using property here? you are just returning `self._kp` and setting it directly in the setter, why not just have `self.kp` as a class attribute directly?

Now, one thing that would be worth doing is have the `kp` property map directly to the `ofc_data` attributes, e.g.:

```
@property
def kp(self) -> float:
    return self.ofc_data.zk_controller["kp"]

@kp.setter
def kp(self, value: float) -> None:
    self.ofc_data.zk_controller["kp"] = value
```

and so on.